edb/pgsql/delta_ext_ai.py

```diff
@@ -0,0 +1,678 @@
+from edb.schema import name as sn
+from edb.schema import properties as s_props
+
+from edb.ir import ast as irast
+
+from edb.edgeql import ast as qlast
+from edb.edgeql import compiler as qlcompiler
+
+from . import codegen
+from . import common
+from . import dbops
+from . import compiler
+from . import types
+from . import ast as pgast
+
+from .common import qname as q
+from .common import quote_literal as ql
+from .common import quote_ident as qi
+from .compiler import astutils
+
+
+ai_index_base_name = sn.QualName("ext::ai", "index")
+
+
+def create_ext_ai_index(
+    index: s_indexes.Index,
+    predicate_src: Optional[str],
+    sql_kwarg_exprs: dict[str, str],
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(schema)
+    assert isinstance(subject, s_indexes.IndexableSubject)
+
+    effective, has_overridden = s_indexes.get_effective_object_index(
+        schema, subject, ai_index_base_name
+    )
+
+    if index != effective:
+        return dbops.CommandGroup()
+
+    # When creating an index on a child that already has an ext::ai index
+    # inherited from the parent, we don't need to create the index, but just
+    # update the populating expressions.
+    if has_overridden:
+        return _refresh_ai_embeddings_source_view(
+            index,
+            options,
+            schema,
+            context,
+        )
+    else:
+        return _create_ai_embeddings(
+            index,
+            predicate_src,
+            sql_kwarg_exprs,
+            options,
+            schema,
+            context,
+        )
+
+
+def delete_ext_ai_index(
+    index: s_indexes.Index,
+    drop_index: dbops.Command,
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+    orig_schema: s_schema.Schema,
+    context: sd.CommandContext,
+) -> dbops.Command:
+    subject = index.get_subject(orig_schema)
+    assert isinstance(subject, s_indexes.IndexableSubject)
+
+    effective, _ = s_indexes.get_effective_object_index(
+        schema, subject, ai_index_base_name
+    )
+
+    if not effective:
+        return _delete_ai_embeddings(index, drop_index, orig_schema, context)
+    else:
+        # effective index remains: don't drop the embeddings
+        return dbops.CommandGroup()
+
+
+def _compile_ai_embeddings_source_view_expr(
+    index: s_indexes.Index,
+    options: qlcompiler.CompilerOptions,
+    schema: s_schema.Schema,
+) -> pgast.SelectStmt:
+    index_sexpr: Optional[s_expr.Expression] = index.get_expr(schema)
+    assert index_sexpr
+    ql = qlast.SelectQuery(
+        result=qlast.Tuple(
+            elements=[
+                qlast.Path(
+                    steps=[qlast.Ptr(name="id")],
+                    partial=True,
+                ),
+                index_sexpr.qlast,
+            ],
+        ),
+    )
+
+    my_options = dataclasses.replace(options, singletons=frozenset())
```

Comment:
Why clearing singletons?

Reply:
What we are doing here is:

```
SELECT (ObjType.id, ObjType."index_expr_of_ext_ai_index")
```

We want all objects, not a single one, but since we're recompiling the `index.expr` we'd better use the options for it.